lib/codecs/src/encoding/format/csv.rs

```diff
@@ -0,0 +1,184 @@
+use crate::encoding::BuildError;
+use bytes::{BufMut, BytesMut};
+use lookup::lookup_v2::ConfigOwnedValuePath;
+use tokio_util::codec::Encoder;
+use vector_core::{
+    config::DataType,
+    event::{Event, Value},
+    schema,
+};
+
+/// Config used to build a `CsvSerializer`.
+#[crate::configurable_component]
+#[derive(Debug, Clone, Default)]
+pub struct CsvSerializerConfig {
+    /// The CSV Serializer Options.
+    pub csv: CsvSerializerOptions,
+}
+
+impl CsvSerializerConfig {
+    /// Creates a new `CsvSerializerConfig`.
+    pub const fn new(csv: CsvSerializerOptions) -> Self {
+        Self { csv }
+    }
+
+    /// Build the `CsvSerializer` from this configuration.
+    pub fn build(&self) -> Result<CsvSerializer, BuildError> {
+        if self.csv.fields.is_empty() {
+            Err("At least one CSV field must be specified".into())
+        } else {
+            Ok(CsvSerializer::new(self.csv.fields.clone()))
+        }
+    }
+
+    /// The data type of events that are accepted by `CsvSerializer`.
+    pub fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    /// The schema required by the serializer.
+    pub fn schema_requirement(&self) -> schema::Requirement {
+        // While technically we support `Value` variants that can't be losslessly serialized to
+        // CSV, we don't want to enforce that limitation to users yet.
+        schema::Requirement::empty()
+    }
+}
+
+/// Config used to build a `CsvSerializer`.
+#[crate::configurable_component]
+#[derive(Debug, Clone, Default)]
+pub struct CsvSerializerOptions {
+    /// The CSV Serializer fields.
+    ///
+    /// The order of the fields determines the order of the columns in the CSV output.
+    /// Make sure to use the same order when parsing the CSV output.
+    ///
+    /// If a field is not present in the event, the output will be `NaN`.
+    ///
+    /// Some value types: Array, Regex, Object are not supported by the CSV format.
+    /// If a field is of an unsupported type, the output will be `NaN`.
+    pub fields: Vec<ConfigOwnedValuePath>,
+}
+
+/// Serializer that converts an `Event` to bytes using the CSV format.
+#[derive(Debug, Clone)]
+pub struct CsvSerializer {
+    fields: Vec<ConfigOwnedValuePath>,
+}
+
+impl CsvSerializer {
+    /// Creates a new `CsvSerializer`.
+    pub const fn new(fields: Vec<ConfigOwnedValuePath>) -> Self {
+        Self { fields }
+    }
+}
+
+impl Encoder<Event> for CsvSerializer {
+    type Error = vector_common::Error;
+
+    fn encode(&mut self, event: Event, buffer: &mut BytesMut) -> Result<(), Self::Error> {
+        let log = event.into_log();
+        let values = log.value();
+        let mut wtr = csv::Writer::from_writer(vec![]);
```

Comment:
You should be able to skip the additional `Vec<u8>` being passed in here, and wrap `Writer` around `buffer` instead, since `BytesMut` implements `io::Write`... and then you'd just have to call `Writer::flush` at the end instead of the current `BufMut::put_slice` call.